lib/src/test/java/io/ably/lib/transport/HostsTest.java

```diff
@@ -0,0 +1,229 @@
+            if (found < 0)
+                allFound = false;
+            else if (found != idx)
+                shuffled = true;
+        }
+        assertThat(idx, is(Defaults.HOST_FALLBACKS.length));
+        assertTrue(allFound);
+        assertTrue(shuffled);
+    }
+
+    /**
+     * Expect a null, when realtimeHost is non-default
+     */
+    @Test
+    public void hosts_fallback_overridden_host() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        String host = "overridden.ably.io";
+        Hosts hosts = new Hosts(host, Defaults.HOST_REALTIME, options);
+        host = hosts.getFallback(host);
+        assertThat(host, nullValue());
+    }
+
+    /**
+     * Expect that returned host is contained within default host list
+     * when realtimeHost is non-default and fallbackHostsUseDefault is set
+     */
+    @Test
+    public void hosts_fallback_use_default() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        options.fallbackHostsUseDefault = true;
+        String host = "overridden.ably.io";
+        Hosts hosts = new Hosts(host, Defaults.HOST_REALTIME, options);
+        int idx;
+        boolean shuffled = false;
+        boolean allFound = true;
+        for (idx = 0; ; ++idx) {
+            host = hosts.getFallback(host);
+            if (host == null)
+                break;
+            int found = Arrays.asList(Defaults.HOST_FALLBACKS).indexOf(host);
+            if (found < 0)
+                allFound = false;
+            else if (found != idx)
+                shuffled = true;
+        }
+        assertThat(idx, is(Defaults.HOST_FALLBACKS.length));
+        assertTrue(allFound);
+        assertTrue(shuffled);
+    }
+
+    /**
+     * It is not allowed to use default fallback hosts and at the same time
+     * provide custom fallback hosts.
+     */
+    @Test(expected = AblyException.class)
+    public void hosts_fallback_use_default_and_set_fallback_hosts() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        options.fallbackHostsUseDefault = true;
+        options.fallbackHosts = new String[] { "custom.ably-realtime.com" };
+
+        new Hosts(null, Defaults.HOST_REALTIME, options);
+    }
+
+    /**
+     * Expect that returned fallback hosts containing the environment information.
+     */
+    @Test
+    public void hosts_fallback_use_environment() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        options.environment = "sandbox";
+        String[] expectedEnvironmentFallbackHosts = Defaults.getEnvironmentFallbackHosts(options.environment);
+        Hosts hosts = new Hosts(null, Defaults.HOST_REALTIME, options);
+
+        String host = "sandbox-" + Defaults.HOST_REALTIME;
+        List<String> returnedEnvironmentFallbackHosts = new ArrayList<>();
+        while ((host = hosts.getFallback(host)) != null){
```

Comment:
It looks to me like `getFallback(host)` with `host` set to `sandbox-<default-host>` as it is here just returns `null`, and therefore `returnedEnvironmentFallbackHosts` remains empty and this test should fail? Perhaps there needs to be an explicit check that it is not empty?

Reply:
Good catch ;) The business logic was updated and this test should be green